src/matrix_utilities.jl

```diff
@@ -6,23 +6,72 @@ function LinearAlgebra.logdet(H::Hermitian)
     return mag + log(resign)
 end
 
+"""
+    qr2chol(pre_array::AbstractMatrix)
+
+Computes the upper triangular cholesky factor of the matrix pre_array'*pre_array
+"""
+function qr2chol(pre_array)
+    right = qr(pre_array).R
+    right_pos = conj.(sign.(Diagonal(right))) * right
+    return UpperTriangular(right_pos)
+end
+
 # matrix square-roots
 lsqrt(m::AbstractMatrix) = cholesky(m).L
 lsqrt(m::UniformScaling) = sqrt(m)
+lsqrt(C::Cholesky) = C.L
 
 # project matrix onto symmetric matrix
 symmetrise(Σ::AbstractMatrix{T}) where {T} = T <: Real ? Symmetric(Σ) : Hermitian(Σ)
 symmetrise(Σ::UniformScaling) = Σ
 symmetrise(Σ::Diagonal) = Σ
+symmetrise(C::Cholesky) = C
+
+"""
+    stein(Σ, Φ, Q)
 
-# stein operator
+Computes the output of the stein  operator
+Σ ↦ Φ*Σ*Φ' + Q
+
+    stein(Σ, Φ)
+
+Same as stein(Σ, Φ, 0.0I)
+"""
 stein(Σ, Φ) = symmetrise(Φ * Σ * Φ')
+stein(Σ::Cholesky, Φ) = Cholesky(qr2chol(Σ.U * Φ'))
+
 stein(Σ, Φ, Q) = symmetrise(Φ * Σ * Φ' + Q)
+stein(Σ::Cholesky, Φ, Q) = Cholesky(qr2chol([Σ.U * Φ'; lsqrt(Q)']))
+stein(Σ::Cholesky, Φ, Q::Diagonal) = Cholesky(qr2chol([Σ.U * Φ'; diagm(sqrt.(Q.diag))]))
```

Comment:
Just wondering since I didn't try the code, but why does the more general function with `lsqrt` not work here?

Reply:
This is kind of annoying but the issue is this: `lsqrt(::Diagonal)` returns `Diagonal` and `vcat(::Matrix,::Diagonal)` returns `SparseArrays.SparseMatrixCSC` which breaks things downstream (It doesnt have multiplication if I recall). 

Though judging by the documentation of `sparse_vcat` in https://docs.julialang.org/en/v1/stdlib/SparseArrays/ this wouldnt be an issue past Julia 1.8.

Reply:
Solution: change compatibility to  1.8 <=